system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,513 @@
+        return MPI::ANY_SOURCE;
+    else
+        return sourceRank;
+}
+
+int getTag(mptag_t mptag)
+{
+    _TF("getTag", mptag);
+    int returnVal;
+    unsigned tag = mptag;
+    if (mptag == TAG_ALL)
+        returnVal = MPI::ANY_TAG;
+    else
+    {
+        if (tag > TAG_UB)
+        {
+            tag= TAG_UB-((unsigned)-tag);                         //MPI doesn't allow custom tags with negative values or beyond MPI_TAG_UB. Thus we shift it to the MPI_TAG_UB range
+        }
+        returnVal = static_cast<int>(tag);
+
+    }
+    return returnVal;
+}
+
+mptag_t getTag(int tag)
+{
+    _TF("getTag", tag);
+    unsigned returnVal;
+    if (tag == MPI::ANY_TAG)
+        returnVal = TAG_ALL;
+    else
+    {
+        if (tag > (TAG_UB/2))
+        {
+            returnVal = (unsigned) - (TAG_UB-tag);                         //MPI doesn't allow custom tags with negative values
+        } else
+            returnVal = tag;
+
+    }
+    return (mptag_t)returnVal;
+}
+
+MPI::Status waitToComplete( bool& completed, bool& error, bool& canceled, bool& timedout, unsigned timeout, CommData *commData)
+{
+    _TF("waitToComplete", completed, error, canceled, timeout);
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+    unsigned remaining;
+    bool noCancellation;
+    while ((noCancellation = commData->lockFromCancellation()) && !(completed || error || (timedout = tm.timedout(&remaining))))
+    {
+        completed = commData->request.Test(stat);
+        commData->releaseCancellationLock();
+        usleep(WAIT_DELAY);
+    }
+    if (completed)
+        canceled = stat.Is_cancelled();
+    else
+        canceled = !noCancellation;
+
+    return stat;
+}
+
+MPI::Status hasIncomingData(int sourceRank, int mptag, MPI::Comm& comm,
+        bool& incomingMessage, bool& error, bool& canceled, unsigned timeout, CommData *commData)
+{
+    _TF("hasIncomingData", sourceRank, mptag, incomingMessage, error, timeout);
+
+    CTimeMon tm(timeout);
+    MPI::Status stat;
+
+    int flag;
+    unsigned remaining;
+    bool noCancellation;
+
+    while ((noCancellation = commData->lockFromCancellation()) && !(incomingMessage || error || (timeout !=0 && tm.timedout(&remaining))))
+    {
+        incomingMessage = comm.Iprobe(sourceRank, mptag, stat);
+        commData->releaseCancellationLock();
+        if (timeout == 0) break;
+        usleep(PROBING_DELAY);
+    }
+    canceled = !noCancellation;
+    return stat;
+}
+
+//----------------------------------------------------------------------------//
+
+/** See mpi_wrapper.hpp header file for function descriptions of the following **/
+
+bool hpcc_mpi::hasIncomingMessage(rank_t &sourceRank, mptag_t &mptag, MPI::Comm& comm)
+{
+    _TF("hasIncomingMessage", sourceRank, mptag);
+    bool incomingMessage = false; bool error = false; bool canceled = false;
+    int source = getRank(sourceRank);
+    int tag = getTag(mptag);
+    CommData* tmpCommData = new CommData(source, tag, comm);
```

Comment:
I'm a bit concerned about thee CommData objects leaking..
Should use linked objects to manage their lifetime and destructions, i.e. this could be a:
```
   Owned<CommData> tmpCommData = new CommData(source, tag, comm);
```